feat: add AI-powered table generation variant (2 of 3)

seanoliver opened the pull request:
## What's this?
Adds AI-powered table generation as the second variant for the [table quickstart experiment](https://www.notion.so/supabase/Table-Editor-Quickstart-2735004b775f80f680eaeb3c6e581633?source=copy_link). This helps new users get started faster by letting them describe their app and having AI generate a complete database schema.

Resolves GROWTH-505

## What changed?

### Architecture: Moved to Empty State (After Design Review)
After discussing with design, we moved the quickstart widget **out of the table editor sidebar** and into the **table editor empty state** (`NewTab.tsx`). This provides a better UX by:
- Making the quickstart more discoverable when users first land on the Table Editor
- Avoiding sidebar clutter and keeping the create table form clean
- Better aligning with the user's mental model (start here → then customize in the form)
- Allowing the widget to be larger and more prominent

The widget appears on the empty state page for projects less than 7 days old, positioned between the "Create a table" action card and recent items.

### AI Table Generation
- Prompt-based schema generation using `gpt-5-mini`
- Generates exactly 3 related tables per prompt
- Streaming responses with AI SDK's `streamObject()` for real-time feedback
- Quick idea buttons for common use cases ("Recipe sharing app", "Event ticketing system", etc.)

### API Implementation (`/api/ai/table-quickstart/generate-schemas`)
- Uses AI SDK's `streamObject()` with Zod schema validation for structured generation
- Streams partial objects as they're generated (tables appear incrementally)
- 50KB buffer limit to prevent memory issues
- Client-side uses `Object.assign()` to merge partial objects efficiently
- Proper abort handling when user navigates away

### State Management
- Extended `onAddTable()` to accept optional `templateData` parameter
- Templates/AI tables flow through the same state mechanism
- When user selects a generated table, it opens the table editor form pre-filled with all fields

## Why these choices?

**Why empty state instead of sidebar?**
Design feedback showed that putting this in the sidebar felt cramped and less discoverable. The empty state is:
- The first thing users see when they land on Table Editor with no tables
- Spacious enough for the widget to breathe
- Naturally positioned as the "starting point" before creating custom tables
- Less intrusive than injecting into the create form

**Why exclude foreign keys?**
The AI can't create meaningful foreign key relationships to tables that don't exist yet. We can add this later if useful, but for now it's cleaner to skip it and let users add relationships manually and use the generated tables as a personalized starting point.

**Why 7-day project gate?**
Only show this to new projects where users are still figuring things out. More established projects probably don't need this kind of hand-holding. Also this is purely an activation experiment for now. If we wanted to build this out into a more fully fledged feature, we'd need to prioritize additional capabilities (foreign keys, multi-table creation, etc.).

**Why 50KB buffer limit?**
Protects against pathological AI responses that could cause memory issues. 50KB is more than enough for 3 table schemas with columns and descriptions. If we hit the limit, we can still use the partial response.

## How to test

### AI Variant (New in this PR)

**Basic Steps**
1. Navigate to the Table Editor page
2. If you have existing tables, delete them all to see the empty state
3. The AI table generator will appear below the "Create a table" button (if project is less than 7 days old)

**Setting the Feature Flag**

To see the AI generator, you need to set the feature flag:
1. Go to PostHog (make sure you're in the staging project)
2. Find the `tableQuickstart` feature flag
3. Click Edit
4. Under "Optional override", select `"ai"`
5. Save the changes

**What to expect**: 
- Input field with quick idea buttons below the "Create a table" card
- Type a prompt (e.g., "fitness tracking app") or click a quick idea
- Hit Generate, tables stream in near-instantly
- Click a table to open the form pre-filled with all fields

### Templates Variant (For validation - already reviewed in #38933)

**Basic Steps**
1. Navigate to the Table Editor empty state
2. The template selector will appear (if project is less than 7 days old)

**Setting the Feature Flag**

To see the template selector:
1. Go to PostHog (staging project)
2. Find the `tableQuickstart` feature flag
3. Set to `"templates"`

**What to expect**: Category tabs at the top, browse templates by category, click to apply.

### Control

To hide the component entirely (baseline for experiment):
1. Set the `tableQuickstart` flag to `"control"` (or disable the flag)

**Note:** The component only shows for projects fewer than 7 days old.

## Notes
This branch contains **both** the templates variant from #38933 and the new AI variant. The templates PR was merged to master first, so this branch has both implementations for easier testing of the experiment variants side-by-side.

Related: Part 3 (#38932) adds telemetry tracking for the experiment.

## Screenshots
`ai` variation on empty state with quick idea suggestions:
<img width="3552" height="1916" alt="CleanShot 2025-10-20 at 20 14 20@2x" src="https://github.com/user-attachments/assets/b92d4535-bc64-4c8f-bb31-08e53bfae4e1" />

`ai` variation results view showing 3 generated tables:
<img width="3540" height="1916" alt="CleanShot 2025-10-20 at 20 15 51@2x" src="https://github.com/user-attachments/assets/80cb458c-df5c-4678-859d-b79e34e0ecba" />

`templates` variation shown in the table editor empty state page:
<img width="3558" height="1930" alt="CleanShot 2025-10-20 at 20 12 54@2x" src="https://github.com/user-attachments/assets/182540d3-0faa-4cf2-ace3-19fc64a67bb5" />

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/useAITableGeneration.ts:
```diff
@@ -0,0 +1,416 @@
+      const response = await fetch(`${BASE_PATH}/api/ai/table-quickstart/generate-schemas`, {
+        method: 'POST',
+        headers,
+        credentials: 'include',
```
Is this necessary? The request shouldn't be cross-origin.

charislam reviewed (changes requested):
It occurs to me now that the "quickstart dismissed" value in local storage should be keyed to project ref, so a user will still see quickstarts for new projects even if they've dismissed it in the past

seanoliver committed:
Refactor and optimize AI table generation

Refactor to use AI SDK streaming with structured object generation:
- Use streamObject() with Zod schema validation for type-safe streaming
- Add 50KB buffer limit to prevent memory issues from oversized responses
- Replace custom stream parsing with Object.assign() for efficient merging
- Use fetchHandler utility for consistent error messages across codebase
- Clean up comments and unused code

seanoliver committed:
Update quickstart widget styling to match action card

- Changed background from bg-surface-75 to bg-surface-100
- Changed hover background to bg-surface-200
- Applied to AI widget, template widget, and quick idea buttons

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/QuickstartAIWidget.tsx:
```diff
@@ -0,0 +1,190 @@
+            type="default"
+            size="tiny"
+            disabled={!aiPrompt.trim() || isGenerating || disabled}
+            onClick={() => handleGenerateTables()}
```
```suggestion
            onClick={handleGenerateTables}
```

seanoliver replied:
ah yeah — i used the inline arrow function here since `handleGenerateTables` takes an optional `tableData` param. if we pass it directly as a function reference, React sends the click event as that first arg and things break.

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/QuickstartAIWidget.tsx:
```diff
@@ -0,0 +1,190 @@
+        </div>
+
+        {apiError && (
+          <div className="text-sm text-red-600" role="alert">
```
We can leave the `role="alert"` as we have no better mechanism to handle this right now, but dynamically adding `role="alert"` typically doesn't work for screen readers -- it's more robust to have a static alert element that's always on the page, and to populate it with new messages as needed

charislam commented on apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/TableEditor/TableQuickstart/useAITableGeneration.ts:
```diff
@@ -0,0 +1,294 @@
+      setIsGenerating(true)
+      setError(null)
+      setTables([])
+    }
```
```suggestion
      setIsGenerating(true)
      setError(null)
      setTables([])
```

same here

seanoliver committed:
Add table quickstart with templates and AI generation

Implement table quickstart feature with two variants:
- Template-based: Pre-built schemas across multiple categories (social, ecommerce, content, productivity)
- AI-powered: Generate custom table schemas from natural language prompts

Key features:
- Template selector with category tabs and schema previews
- AI generation using gpt-oss-120b via Bedrock for fast response times
- Streaming responses with real-time table generation
- Quick idea chips for common use cases
- Success feedback and seamless integration with table editor
- Feature flag gated (tableQuickstart) with 7-day new project window
- Comprehensive error handling and retry logic
- ARIA labels and accessibility improvements

Technical implementation:
- Custom React hooks for AI generation and state management
- Zod schema validation for type-safe streaming
- API endpoint with structured object streaming
- localStorage-based dismissal state
- Consistent styling with existing UI patterns

seanoliver commented:
> Oh, another thing I noted when testing:
> 
> * If the AI generates multiple tables, and you create one
> * You then lose the others, because it immediately opens the newly created table in a new tab
> * And even if you close that tab, the old AI suggestions are lost so you have to regenerate from scratch


Great point! I've updated it to persist the last generation in localStorage so users can create multiple tables from the same results and even come back later to find them still there. They only get cleared when you click "Generate new" or do a new search.

seanoliver committed:
Address PR review feedback

- Persist AI results to localStorage so users can create multiple tables from same generation
- Remove unnecessary isMountedRef checks before synchronous state updates
- Change Quick ideas label to semantic h4 element for accessibility
- Add TODO comment about migrating to experimental_useObject() hook

dnywh reviewed (approved):
Smoke tested both `ai` and templates` variants locally. I love this.

Will leave minor (non-blocking) feedback elsewhere.

seanoliver committed:
Clean up table quickstart implementation

- Match SQL Editor ActionCard styling with circular icons
- Implement smart Generate/Regenerate button logic
- Add tooltips for column counts
- Persist prompt to localStorage
- Fix React hook dependency warnings
- Remove debugging artifacts
